### PR TITLE
Implement legacy DAR support, #308

### PR DESCRIPTION
**Pull request by leo-da**
if MANIFEST.MF is not present or does not contain required fields, filter zip entries by `.dalf` extension,
expect one `main` DALF and one optional `*-prim*` DALF, fail in other cases

Closes: #308

### CODE REVIEW

**Comment by leo-da** [comment]
there is no test case for the fall-back scenario. However I tested it manually. Need to think how we can generate a legacy DAR.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala [hunk 1]
```diff
@@ -22,15 +24,40 @@ class DarReader[A](
...
+    }
+
   private def parseDalfNamesFromManifest(darFile: ZipFile): Try[Dar[String]] =
     bracket(Try(darFile.getInputStream(manifestEntry)))(close)
       .flatMap(is => readDalfNamesFromManifest(is))
 
+  private def findDalfNames(darFile: ZipFile): Try[Dar[String]] = {
+    val entries: List[ZipEntry] = darFile.entries.asScala.toList
+    val dalfs: List[String] = entries.filter(isDalf).map(_.getName)
+    dalfs.partition(isPrimDalf) match {
```
Comment (gerolf-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala]: This won't work for DARs that also contain the stdlib, which is now produced as a separate DALF file next to `daml-prim.dalf` and the DALF file for the actual model. Unless we never want to support a DAR file with no manifest but more than 2 DALF files.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala [hunk 2]
```diff
@@ -46,6 +73,17 @@ class DarReader[A](
   private def close(is: InputStream): Try[Unit] = Try(is.close())
 }
 
+case class InvalidLegacyDar(dalfNames: List[String])
+    extends RuntimeException(
+      s"Invalid Legacy DAR. Legacy DAR can contain: one main DALF and an optional *-prim* DALF. " +
+        s"This DAR contains: [${dalfNames.mkString(", "): String}]")
+
+case class InvalidDar(cause1: Throwable, cause2: Throwable)
```
Comment (gerolf-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala]: I wouldn't mention the legacy mode at all. If both "modes" fail to read the DAR file, we should just fail with the error message of the currently expected format (aka DAR with manifest). For most users the legacy mode should never come into action.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala (hunk 1 above)
Comment (francesco-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala]: > Unless we never want to support a DAR file with no manifest but more than 2 DALF files.
that is my assumption here. @robin-da can you comment?

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala (hunk 2 above)
Comment (francesco-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala]: i'm happy for @leo-da to adjust the error message, i'll merge this in the meantime.

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala (hunk 1 above)
Comment (leo-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala]: @francesco-da legacy DAR only contained main and prim dalfs, no stdlib. But again, specification for the DAR format would be nice to have

Comment (leo-da) [reply daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala]:
```
leo   [42 minutes ago]
Yep, that's what I remember. The old dar that didn't have Main-Dalf in the manifest could only have two dalfs, main and prim. So no reason to overcomplicate the dar reader. New dars with stdlib will always have Main-Dalf in the manifest. Am I right? @robin
And yeah we need a spec, if format keeps changing (edited)


robin   [36 minutes ago]
yes
```

**Pull request merged**